benches/common/supernova/bench.rs

```diff
@@ -0,0 +1,179 @@
+
+      recursive_snark
+        .verify(&pp, &z0_primary, &z0_secondary)
+        .expect("Verify failed");
+    } else {
+      unimplemented!()
+    }
+
+    selected_augmented_circuit = (selected_augmented_circuit + 1) % num_augmented_circuits;
+    recursive_snark_option = Some(recursive_snark)
+  }
+
+  assert!(recursive_snark_option.is_some());
+  let recursive_snark = recursive_snark_option.unwrap();
+
+  let bench_params = BenchParams {
+    step_size: num_cons,
+    date: env!("VERGEN_GIT_COMMIT_DATE"),
+    sha: env!("VERGEN_GIT_SHA"),
+  };
+
+  match snark_type {
+    SnarkType::Compressed => {
+      let (prover_key, verifier_key) = CompressedSNARK::<_, _, _, _, S1, S2>::setup(&pp).unwrap();
+      // Benchmark the prove time
+      group.bench_function(bench_params.bench_id("Prove"), |b| {
+        b.iter(|| {
+          assert!(CompressedSNARK::<_, _, _, _, S1, S2>::prove(
+            black_box(&pp),
+            black_box(&prover_key),
+            black_box(&recursive_snark)
+          )
+          .is_ok());
+        })
+      });
+
+      let res = CompressedSNARK::<_, _, _, _, S1, S2>::prove(&pp, &prover_key, &recursive_snark);
+      assert!(res.is_ok());
+      let compressed_snark = res.unwrap();
+      // Benchmark the verification time
+      group.bench_function(bench_params.bench_id("Verify"), |b| {
+        b.iter(|| {
+          assert!(black_box(&compressed_snark)
+            .verify(
+              black_box(&pp),
+              black_box(&verifier_key),
+              black_box(&z0_primary),
+              black_box(&z0_secondary),
+            )
+            .is_ok());
+        })
+      });
+    }
+    SnarkType::Recursive => {
+      // Benchmark the prove time
+      group.bench_function(bench_params.bench_id("Prove"), |b| {
+        b.iter(|| {
+          assert!(black_box(&mut recursive_snark.clone())
+            .prove_step(
+              black_box(&pp),
+              &bench.primary_circuit(0),
+              &bench.secondary_circuit()
+            )
+            .is_ok());
+        })
+      });
+
+      // Benchmark the verification time
+      group.bench_function(bench_params.bench_id("Verify"), |b| {
+        b.iter(|| {
+          assert!(black_box(&mut recursive_snark.clone())
+            .verify(
+              black_box(&pp),
+              black_box(&[<PallasEngine as Engine>::Scalar::from(2u64)]),
+              black_box(&[<VestaEngine as Engine>::Scalar::from(2u64)]),
+            )
+            .is_ok());
+        })
+      });
+    }
+  }
+}
+
+pub fn run_bench<S1: BatchedRelaxedR1CSSNARKTrait<E1>, S2: RelaxedR1CSSNARKTrait<E2>>(
+  c: &mut Criterion,
+  group_name: &str,
+  arity: usize,
+  snark_type: SnarkType,
+) {
+  // we vary the number of constraints in the step circuit
+  for &num_cons_in_augmented_circuit in num_cons().iter() {
+    // number of constraints in the step circuit
+    let num_cons = num_cons_in_augmented_circuit - NUM_CONS_VERIFIER_CIRCUIT_PRIMARY;
```

Comment:
Yeah, so this is the problem: in bench mode, that substraction has wrapping semantics, and if a missed update of the cons numbers makes it wrap around, the benches will silently start eating up the bench machine's memory. We should do `checked_sub(..).expect("constraint numbers out of date!!")` here